x/cdp/keeper/auctions.go

```diff
@@ -38,29 +43,53 @@ func (pd partialDeposits) SumDebt() (sum sdk.Int) {
 }
 
 // AuctionCollateral creates auctions from the input deposits which attempt to raise the corresponding amount of debt
-func (k Keeper) AuctionCollateral(ctx sdk.Context, deposits types.Deposits, debt sdk.Int, bidDenom string) {
+func (k Keeper) AuctionCollateral(ctx sdk.Context, deposits types.Deposits, debt sdk.Int, bidDenom string) sdk.Error {
 	auctionSize := k.getAuctionSize(ctx, deposits[0].Amount[0].Denom)
 	partialAuctionDeposits := partialDeposits{}
 	totalCollateral := deposits.SumCollateral()
-	for deposits.SumCollateral().GTE(auctionSize) {
+	for totalCollateral.GT(sdk.ZeroInt()) {
 		for i, dep := range deposits {
 			// create auctions from individual deposits that are larger than the auction size
-			k.CreateAuctionsFromDeposit(ctx, &dep, &debt, &totalCollateral, auctionSize, bidDenom)
-			if !dep.Amount[0].Amount.IsZero() {
-				debtCoveredByDeposit := (dep.Amount[0].Amount.Quo(totalCollateral)).Mul(debt)
-				if (partialAuctionDeposits.SumCollateral().Add(dep.Amount[0].Amount)).LT(auctionSize) {
+			debtChange, collateralChange, err := k.CreateAuctionsFromDeposit(ctx, dep, debt, totalCollateral, auctionSize, bidDenom)
+			if err != nil {
+				return err
+			}
+			debt = debt.Sub(debtChange)
+			totalCollateral = totalCollateral.Sub(collateralChange)
+			dep.Amount = sdk.NewCoins(sdk.NewCoin(dep.Amount[0].Denom, dep.Amount[0].Amount.Sub(collateralChange)))
+			// if there is leftover collateral that is less than a lot
+			if !dep.Amount.IsZero() {
+				collateralAmount := dep.Amount[0].Amount
+				collateralDenom := dep.Amount[0].Denom
```

Comment:
Can move these up a couple of lines and replace `dep.Amount[0].Denom`, `dep.Amount[0].Amount` on line 59 with the local variables.